buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -1052,179 +1257,389 @@ private void transferToNextInLineOrRelease(Chunk chunk) {
             chunk.releaseFromMagazine();
         }
 
-        boolean trySetNextInLine(Chunk chunk) {
-            return NEXT_IN_LINE.compareAndSet(this, null, chunk);
-        }
-
+        @Override
         void free() {
             // Release the current Chunk and the next that was stored for later usage.
             restoreMagazineFreed();
-            long stamp = allocationLock != null ? allocationLock.writeLock() : 0;
+            long stamp = allocationLock.writeLock();
             try {
                 if (current != null) {
                     current.releaseFromMagazine();
                     current = null;
                 }
             } finally {
-                if (allocationLock != null) {
-                    allocationLock.unlockWrite(stamp);
-                }
-            }
-        }
-
-        public AdaptiveByteBuf newBuffer() {
-            AdaptiveByteBuf buf;
-            if (handle == null) {
-                buf = EVENT_LOOP_LOCAL_BUFFER_POOL.get();
-            } else {
-                buf = bufferQueue.poll();
-                if (buf == null) {
-                    buf = new AdaptiveByteBuf(handle);
-                }
+                allocationLock.unlockWrite(stamp);
             }
-            buf.resetRefCnt();
-            buf.discardMarks();
-            return buf;
-        }
-
-        boolean offerToQueue(Chunk chunk) {
-            return group.offerToQueue(chunk);
-        }
-
-        public void initializeSharedStateIn(Magazine other) {
-            chunkController.initializeSharedStateIn(other.chunkController);
         }
     }
 
-    private static final class ChunkRegistry {
-        private final LongAdder totalCapacity = new LongAdder();
+    private static final class ThreadLocalMagazine extends AbstractMagazine {
+        private static final AtomicLongFieldUpdater<ThreadLocalMagazine> EXTERNAL_USED_MEMORY_UPDATER =
+                AtomicLongFieldUpdater.newUpdater(ThreadLocalMagazine.class, "externalUsedMemory");
+        private final ArrayDeque<AdaptiveByteBuf> localBuffers;
+        private volatile long externalUsedMemory; // For non-owner threads
+        private long localUsedMemory; // For owner threads
```

Comment:
you still need a volatile here and an updater: once you need an external thread to read the total used memory and you have to read this value it still needs to be "published" correctly (i.e. atomically).
IN order to achieve it you can use `LOCALL_USED_MEMORY_UPDATER.lazySet(this, localUsedMemory + value)` or you can even push this forward (in case it matters on Apple MX) as:

```java
        private volatile long externalUsedMemory; // For non-owner threads
        private volatile long localUsedMemoryView; // For non-owner threads, updated via lazySet(localUsedMemory)
        private long localUsedMemory; // incremented without any need to perform volatile loads/store nor lazy set - just pain +=
```